test/envtest/gateway_secret_watch_envtest_test.go

```diff
@@ -0,0 +1,155 @@
+package envtest
+
+import (
+	"context"
+	"testing"
+
+	"github.com/samber/lo"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gatewayv1 "sigs.k8s.io/gateway-api/apis/v1"
+
+	kogateway "github.com/kong/kong-operator/controller/gateway"
+	certhelper "github.com/kong/kong-operator/ingress-controller/test/helpers/certificate"
+	gwcdecor "github.com/kong/kong-operator/internal/utils/gatewayclass"
+	"github.com/kong/kong-operator/modules/manager/logging"
+	managerscheme "github.com/kong/kong-operator/modules/manager/scheme"
+	k8sutils "github.com/kong/kong-operator/pkg/utils/kubernetes"
+	"github.com/kong/kong-operator/pkg/vars"
+)
+
+func TestGatewaySecretWatch_UpdatesResolvedRefsOnSecretRotation(t *testing.T) {
+	t.Parallel()
+
+	// Prepare scheme, envtest, manager and KO Gateway reconciler.
+	scheme := managerscheme.Get()
+	ctx := t.Context()
+
+	cfg, ns := Setup(t, ctx, scheme)
+	mgr, logs := NewManager(t, ctx, cfg, scheme)
+
+	r := &kogateway.Reconciler{
+		Client:                mgr.GetClient(),
+		Scheme:                scheme,
+		Namespace:             ns.Name,
+		DefaultDataPlaneImage: "kong:latest",
+		LoggingMode:           logging.DevelopmentMode,
+	}
+	StartReconcilers(ctx, t, mgr, logs, r)
+
+	c := mgr.GetClient()
+
+	// Create a GatewayClass accepted by the controller.
+	gc := &gatewayv1.GatewayClass{
+		ObjectMeta: metav1.ObjectMeta{Name: "gc-ko"},
+		Spec: gatewayv1.GatewayClassSpec{
+			ControllerName: gatewayv1.GatewayController(vars.ControllerName()),
+		},
+	}
+	require.NoError(t, c.Create(ctx, gc))
+
+	t.Log("patching GatewayClass status to Accepted=True")
+	require.Eventually(t, func() bool {
+		var cur gatewayv1.GatewayClass
+		if err := c.Get(ctx, types.NamespacedName{Name: gc.Name}, &cur); err != nil {
+			return false
+		}
+		cond := metav1.Condition{
+			Type:               string(gatewayv1.GatewayClassConditionStatusAccepted),
+			Status:             metav1.ConditionTrue,
+			Reason:             string(gatewayv1.GatewayClassReasonAccepted),
+			ObservedGeneration: cur.Generation,
+			LastTransitionTime: metav1.Now(),
+		}
+		// Use shared helper to set/merge the condition.
+		gwcd := gwcdecor.DecorateGatewayClass(&cur)
+		k8sutils.SetCondition(cond, gwcd)
+		if err := c.Status().Update(ctx, &cur); err != nil {
+			return false
+		}
+		return true
```

Comment:
This looks like something that we could reuse. We already have a bunch of similar helpers in https://github.com/Kong/kong-operator/blob/e90fd984547cd60e3ce8c481f368514dd1936411/pkg/utils/test/predicates.go, e.g. https://github.com/Kong/kong-operator/blob/e90fd984547cd60e3ce8c481f368514dd1936411/pkg/utils/test/predicates.go#L463-L484.

I think it would make sense to put it there and use it here. WDYT?